src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs

```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
+                        for (int i = 0; i < len; ++i)
+                        {
+                            arg.ArrayValue[i] = ParseCustomAttributeValue(ref parser, attributeType, module);
+                        }
+                    }
+                    break;
+                }
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+
+            return arg;
+        }
+
+        private static CustomAttributeType ParseCustomAttributeType(ref CustomAttributeDataParser parser, RuntimeModule module)
+        {
+            CustomAttributeEncoding arrayTag = CustomAttributeEncoding.Undefined;
+            CustomAttributeEncoding enumTag = CustomAttributeEncoding.Undefined;
+            Type? enumType = null;
+
+            CustomAttributeEncoding tag = parser.GetTag();
+            if (tag is CustomAttributeEncoding.Array)
+            {
+                arrayTag = parser.GetTag();
+            }
+
+            // Load the enum type if needed.
+            if (tag is CustomAttributeEncoding.Enum
+                || (tag is CustomAttributeEncoding.Array
+                    && arrayTag is CustomAttributeEncoding.Enum))
+            {
+                // We cannot determine the underlying type without loading the enum.
+                string? enumTypeMaybe = parser.GetString();
+                if (enumTypeMaybe is null)
+                {
+                    throw new ArgumentNullException();
+                }
+
+                enumType = TypeNameParser.GetTypeReferencedByCustomAttribute(enumTypeMaybe, module);
+                if (enumType is null || !enumType.IsEnum)
+                {
+                    throw new ArgumentOutOfRangeException();
+                }
+
+                enumTag = RuntimeCustomAttributeData.TypeToCustomAttributeEncoding((RuntimeType)enumType.GetEnumUnderlyingType());
+            }
+
+            return new CustomAttributeType(tag, arrayTag, enumTag, enumType);
+        }
+
+        /// <summary>
+        /// Used to parse CustomAttribute data. See ECMA-335 II.23.3.
+        /// </summary>
+        private ref struct CustomAttributeDataParser
         {
-            m_type = type;
-            m_encodedArgument = default;
+            private int _curr;
+            private ReadOnlySpan<byte> _blob;
+
+            public CustomAttributeDataParser(ConstArray attributeBlob)
+            {
+                unsafe
+                {
+                    _blob = new ReadOnlySpan<byte>((void*)attributeBlob.Signature, attributeBlob.Length);
+                }
+                _curr = 0;
+            }
+
+            private ReadOnlySpan<byte> PeekData(int size)
+            {
+                ArgumentOutOfRangeException.ThrowIfLessThan(_blob.Length - _curr, size);
+                return _blob.Slice(_curr, size);
+            }
+
+            private ReadOnlySpan<byte> ReadData(int size)
+            {
+                var tmp = PeekData(size);
```

Comment:
```suggestion
                ReadOnlySpan<byte> tmp = PeekData(size);
```
This is invalid use of `var` per https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md